### ISSUE TITLE
MissingMethodException: Method not found: 'Void Microsoft.AspNetCore.Builder.IApplicationBuilder.set_ApplicationServices

### ISSUE THREAD

**Issue by MichalStrehovsky** [dotnet/linker#2820]
I wanted to confirm a trimming issue in NativeAOT is caused by a known NativeAOT bug (treating TrimmerDefaultAction=copy same as TrimmerDefaultAction=copyused) but instead of confirming it, I found an illink bug:

```
C:\illinkrepro>dotnet --info
.NET SDK:
 Version:   7.0.100-preview.6.22302.2
 Commit:    b42b660f0b
```

```
dotnet new webapi -o illinkrepro
cd illinkrepro
dotnet publish -r win-x64 -c Release /p:PublishTrimmed=true /p:TrimmerDefaultAction=copyused

C:\illinkrepro>C:\illinkrepro\bin\Release\net7.0\win-x64\publish\illinkrepro.exe
Unhandled exception. System.MissingMethodException: Method not found: 'Void Microsoft.AspNetCore.Builder.IApplicationBuilder.set_ApplicationServices(System.IServiceProvider)'.
   at Program.<Main>$(String[] args)
```

**Comment by davidfowl** [comment]
Woah, can you send me the repro for this one?

**Comment by MichalStrehovsky** [comment]
> 

The repro is in the top-post - just dotnet publish the default webapi template with `/p:PublishTrimmed=true /p:TrimmerDefaultAction=copyused`. Note I believe this is an illink bug - we hit MissingMethodException in Main which usually means illink stripped a method `Main` refers to (or a method we need to execute Main) - it doesn't look reflection related.

**Comment by marek-safar** [comment]
Does it happen `copyused` only?

**Comment by MichalStrehovsky** [comment]
> Does it happen `copyused` only?

Yup. `copy` works, and `link` runs into `Could not load file or assembly 'Microsoft.AspNetCore.OpenApi'` that is hit because the code is trimming-unfriendly (not an illink issue).

**Comment by vitek-karas** [comment]
We should definitely understand what's going on - that said `copyused` is something we're trying to eventually remove.
@MichalStrehovsky  is this something you were just trying out, or is it because of some specific scenario/issue?

**Comment by MichalStrehovsky** [comment]
> @MichalStrehovsky is this something you were just trying out

I wanted to see if some behavior is a NativeAOT bug or trim-hostiliy of the app. Running illink in the same config as NativeAOT would quickly help narrow that down, so that's what I tried. I'm not blocked on this fix.

**Comment by DamianEdwards** [comment]
Hitting this in my ["try out trimming repo"](https://github.com/DamianEdwards/TrimmedTodo) too. If I try to enable `<TrimmerDefaultAction>copyused</TrimmerDefaultAction>` in the web projects, app startup fails with:

```
Unhandled exception. System.MissingMethodException: Attempted to access a missing method.
   at Program.<Main>$(String[] args)
   at Program.<Main>(String[] args)
```
